factor/u2f/classes/factor.php

```diff
@@ -0,0 +1,251 @@
+        global $USER, $CFG, $DB;
+        $return = array();
+        if(empty($data['response_input'])) {
+            $return['verificationcode'] = get_string('error', 'factor_u2f');
+            return $return;
+        }
+
+        $factors = $this->get_all_user_factors($USER);
+        $registrations = [];
+        foreach ($factors as $f) {
+            $registrations[$f->id] = json_decode($f->secret);
+        }
+        $url = parse_url($CFG->wwwroot);
+        $u2f = new U2F($url['scheme'].'://'.$url['host']);
+        try {
+            $authentication = $u2f->doAuthenticate(json_decode($data['request']), $registrations, json_decode($data['response_input']));
+            foreach ($registrations as $id => $registration) {
+                if ($authentication->keyHandle === $registration->keyHandle) {
+                    $row = $DB->get_record('tool_mfa', array('id' => $id));
+                    $row->secret = json_encode($authentication);
+                    $row->timemodified = time();
+                    $row->lastverified = time();
+
+                    $DB->update_record('tool_mfa', $row);
+
+                    break;
+                }
+            }
+        } catch (Error $e) {
+            $return['verificationcode'] = get_string('error', 'factor_u2f'); //TODO
+        }
+
+
+        return $return;
+    }
+
+    /**
+     * U2F Factor implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function get_all_user_factors($user) {
+        global $DB;
+
+        $records = $DB->get_records('tool_mfa', array(
+            'userid' => $user->id,
+            'factor' => $this->name,
+        ));
+        return $records;
+    }
+
+    /**
+     * Users need to set up their own credentials by registering their individual token(s).
+     * @return bool
+     */
+    public function has_setup() {
+        return true;
+    }
+
+    /**
+     * Users are able to revoke individual U2F factors.
+     * @return bool
+     */
+    public function has_revoke() {
+        return true;
+    }
+
+    /**
+     * U2F Factor implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function get_state() {
+        global $USER;
+        $userfactors = $this->get_active_user_factors($USER);
+
+        // If no codes are setup then we must be neutral not unknown.
+        if (count($userfactors) == 0) {
+            return \tool_mfa\plugininfo\factor::STATE_NEUTRAL;
+        }
+
+        return parent::get_state();
+    }
+
+    /**
+     * U2F Factor implementation.
+     *
+     * @param $mform
+     */
+    public function setup_factor_form_definition($mform) {
+        global $PAGE, $CFG;
+
+
+
+        // Prepare data for U2F request.
+        $url = parse_url($CFG->wwwroot);
+
+        if($url['scheme'] == "http"){
+            $mform->addElement('html', '<div class="m-element-notification m-element-notification--warning">
+            You cannot use https for u2f authentication.</div></div>');
```

Comment:
typo, http vs https. Also I would use `$OUTPUT->notification('...', \core\output\notification::NOTIFY_ERROR)` instead of the custom html

Reply:
actually I don't think we should do this check at all. Yes I totally agree that sites should be on https, but this isn't the place to enforce it and there is no technical reason to disallow things here.

Reply:
I would definitely agree that this is not the place to perform a check here. However, the thing is that there actually is a technical reason to disallow HTTP here: The U2F protocol refuses to work if the connection is established via HTTP. So this is a hard fail here. And I would keep this check (and notification) in place, primarily because it would make any developer on `http://localhost/` tear their hair out ;)

Removing the custom HTML though, of course.